torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2247,88 +2385,220 @@ def _load_sharded_state_dict(
-            # p._local_shard, which would reside on CPU.
-            self._use_param_local_shard()
-
-            # Register pre-backward hooks to all-gather the params for the backward
-            # pass (if output's grad was needed). This won't register anything if
-            # we are in eval mode.
-            outputs = self._register_pre_backward_hooks(outputs)
-
-            # Done with a forward pass.
+            if reshard_fn is not None:
+                reshard_fn()
+            # Register pre-backward hooks to unshard the flattened parameters for
+            # the gradient computation if needed
+            output = self._register_pre_backward_hooks(output, [handle.flat_param for handle in handles])
             self.training_state = TrainingState_.IDLE
+            return output
 
-        return outputs
+    def _cast_forward_inputs(self, *args, **kwargs):
+        """
+        Moves the :meth:`forward` inputs to the compute device and casts them
+        to the appropriate dtype if needed.
+        """
+        if not self._is_root:
+            return
+        # TODO: Do not use the side stream for tensor copies for now;
+        # investigate the perf with/without it
+        # TODO: For mixed precision, move the inputs to the compute device and
+        # cast to reduced-precision in a single `to()` call
+        args, kwargs = _to_kwargs(args, kwargs, self.compute_device.index, False)
+        args = args[0]
+        kwargs = kwargs[0]
+        if self._mixed_precision_enabled_for_params():
+            input_dtype = self.mixed_precision.param_dtype
+            args, kwargs = self._cast_fp_inputs_to_precision(
+                input_dtype, *args, **kwargs,
+            )
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        """
+        Runs the forward pass. For the recursive-wrapping path, the pre-forward
+        and post-forward are called explicitly, while for the non-recursive-
+        wrapping path, they are registered as hooks on every (sub)module.
+        """
+        with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):    
+            # Non-recursive wrapping path (using hooks)
+            if self._use_param_exec_order_policy:
+                self._cast_forward_inputs(*args, **kwargs)
+                return self._fsdp_wrapped_module(*args, **kwargs)
+            # Recursive wrapping path
+            unused = None
+            free_full_params = self.reshard_after_forward
+            free_mp_shard = (
+                self.reshard_after_forward and
+                self._mixed_precision_enabled_for_params()
+            )
+            reshard_fn = functools.partial(
+                self._reshard,
+                [handle.flat_param for handle in self._handles],
+                free_full_params,
+                free_mp_shard,
+            )
+            self._pre_forward(self._handles, unused, unused, unused)
```

Comment:
so for the non-recursive root module: I assume here it will rebuild all params since `self._handles` is passed into `_pre_forward`?

Reply:
I think this is the recursive path. It will rebuild all of the calling FSDP instance's handles, which is just one or zero handles in our current design.

Reply:
so where would the forward of the non-recursive FSDP root module be called?

Reply:
It is the above:
```
if self._use_param_exec_order_policy:
    self._cast_forward_inputs(*args, **kwargs)
    return self._fsdp_wrapped_module(*args, **kwargs)
```
Once it enters `self._fsdp_wrapped_module(*args, **kwargs)`, it begins running the `forward()` of each module in the wrapped model, where each module has pre and post-forward hooks registered.

Reply:
Oh I see I see, this makes sense